plugins/com.google.cloud.tools.eclipse.appengine.facets/src/com/google/cloud/tools/eclipse/appengine/facets/StandardFacetInstallDelegate.java

```diff
@@ -75,6 +75,7 @@ private AppEngineRuntimeInstallJob(IFacetedProject facetedProject) {
       super(Messages.getString("appengine.install.runtime.to.project", // $NON-NLS$
           facetedProject.getProject().getName()));
       this.facetedProject = facetedProject;
+      setRule(facetedProject.getProject());
```

Comment:
Ah, the deadlock resolution is because I added a time limit when waiting for the first `ConvertJob` with `MAX_JSDT_CHECK_RETRIES`:

```java
        IProjectFacet jsdtFacet = ProjectFacetsManager.getProjectFacet(JSDT_FACET_ID);
        for (int times = 0;
            times < MAX_JSDT_CHECK_RETRIES && !facetedProject.isFixedProjectFacet(jsdtFacet);
            times++) {
 ```

After 10 seconds, it just gives up. So, this is a deadlock that is always reproducible.